src/Magnum/Audio/Listener.h

```diff
@@ -0,0 +1,247 @@
+
+namespace Magnum { namespace Audio {
+
+/** !!! TODO: This is an experiment !!!
+ * @brief Base for @ref AbstractListener, @ref Listener2D and @ref Listener3D
+ */
+class AbstractListenerBase {
+    public:
+
+        /** @brief Destructor */
+        virtual ~AbstractListenerBase() = 0;
+
+        /** @brief Whether this listener is the active listener */
+        bool isActive() {
+            return (this == _activeListener);
+        }
+
+        /**
+         * @brief Set this listener as the active listener
+         * @param active Whether to activate or deactivate this listener.
+         *
+         * When activating a listener, all other listeners are disabled. If
+         * this listener was not active and is disabled by passing `false`,
+         * this method does nothing.
+         */
+        virtual void setActive(bool active = true) {
+            if(active) {
+                _activeListener = this;
+            } else if(isActive()){
+                _activeListener = nullptr;
+            }
+        }
+
+    protected:
+        /** @brief Constructor */
+        explicit AbstractListenerBase() = default;
+
+    private:
+        /* Pointer to the currently active listener. Should never be dereferenced, since may be null at any given time. */
+        static AbstractListenerBase* _activeListener;
+};
+
+AbstractListenerBase* AbstractListenerBase::_activeListener = nullptr;
+
+inline AbstractListenerBase::~AbstractListenerBase() { }
+
+/**
+@brief Base of @ref Listener2D and @ref Listener3D
+
+Contains methods common for all Listeners.
+
+## Usage
+
+Minimal scene setup with a listener will require the following code:
+
+@code
+Scene3D scene;
+Object3D object{&scene};
+Listener3D listener{object};
+
+// Important!
+listener.setActive();
+
+// ... and every frame, update listener to object transformation if necessary
+listener.update();
+
+@endcode
+
+Without calling `listener.setActive()`, calling `listener.update()` will not update the OpenAL listener.
+For two dimensional scenes simply replace `3D` with `2D`.
+
+## Active Listener
+
+There can only be at the most *one* active listener at a given time. This is because OpenAL only supports notion of one listener.
+Having multiple @ref Listener2D or @ref Listener3D instances can still be useful for conveniently switching between them for
+cinematics for example.
+
+## Sound Transformation
+
+@ref AbstractListener::setSoundTransformation() enables you to set transformation which is applied to the listeners orientation and position
+before passed onto OpenAL. This can be used for rotating two dimensional scenes as planes into three dimensional space or even scaling the audio
+scene to match a certain world scale. In the later case you might want to instead consider @ref Renderer::setSpeedOfSound().
+
+*/
+template <UnsignedInt dimensions, class Derived> class AbstractListener: public SceneGraph::AbstractFeature<dimensions, Float>, public AbstractListenerBase {
```

Comment:
Right. Then I think it would make sense to combine all classes (`AbstractListener`, `Listener2D` and `Listener3D`) into one template (along with removal of `setRotationAxis()`). The amount of generated code would be the same as in current case.